### PR TITLE
Change auto_cast in SuperVectorizer

### PR DESCRIPTION
**Pull request by LeoGrin**
Uses pandas [`TextParser`](https://github.com/pandas-dev/pandas/blob/06d230151e6f18fdb8139d09abf539867a8cd481/pandas/io/parsers/readers.py#L1289) (used in `read_csv`, as per @GaelVaroquaux suggestion in #234) instead of `convert_dtypes` in the SuperVectorizer `_auto_cast` method. This has several advantages:

- Type detection works on string columns, e.g `["3", "2", "1"]` would get detected as integer. Solves #234 
-  Missing values encoded as strings get treated as missing values, and don't prevent type detection. For instance,` [1, 2, "", 4]` is detected as integers. (solves  #231). Which strings get detected as Nans is easy to change (see below).
- Datetime columns are converted to datetimes, and the format is inferred by Pandas. Solve the first part of #233.

Some comments:
- TODO: I'd like to enable the user to specify a specific date format if he wants to.
- Strings detected as NaNs are the default [`STR_NA_VALUES`](https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/parsers.pyx#L1350), and I've added `[None, " ", "?", "..."]`. I think we could find a more principled way to choose which strings to include.
- I haven't properly measured the speed of the conversion.
- To use TextParser, I need to convert the input data into a list of list. I don't think this is a problem, but tell me if you think otherwise.
- All variables which are not numerics or datetimes are converted to object (and thus handled as categorical in the SuperVectorizer).
- Right now, if a datetime column contains mixed timezones, it is encoded as an object.

### CODE REVIEW

File: dirty_cat/super_vectorizer.py
Comment (LilianBoulard) [inline dirty_cat/super_vectorizer.py:183]:
```suggestion
        pd.DataFrame
```

File: dirty_cat/super_vectorizer.py
Comment (LilianBoulard) [inline dirty_cat/super_vectorizer.py:184]:
```suggestion
            The same array as a pandas DataFrame, with its columns casted to the best possible
```

File: dirty_cat/super_vectorizer.py
Comment (LilianBoulard) [inline dirty_cat/super_vectorizer.py:203]: Should it be useful to account for columns with type `int32` or similar ?
Somewhere in the code, I've used [is_numeric_dtype](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.api.types.is_numeric_dtype.html), perhaps it could also make the code easier to read ?

File: dirty_cat/super_vectorizer.py
Comment (LilianBoulard) [inline dirty_cat/super_vectorizer.py:261]: Same comment as above

**Review by LilianBoulard — COMMENTED** [review]
Thanks for the PR, it looks nice !

**Review by LilianBoulard — APPROVED** [review]
LGTM !

File: dirty_cat/super_vectorizer.py
Comment (GaelVaroquaux) [inline dirty_cat/super_vectorizer.py:192]: I fear that on large arrays, this will blow the memory. Can we do type detection only on the first few thousand lines?

**Comment by LeoGrin** [comment]
In trying to take into account @GaelVaroquaux's [comment](https://github.com/dirty-cat/dirty_cat/pull/238#discussion_r810191767), I've actually realized that instead of using TextParser, it would be simpler to go back to @LilianBoulard's version of autocast, and:
- replace convert_dtypes by pd.to_numeric and pd.to_datetime
- Replace every missing value with np.nan *before* type conversion

**Comment by GaelVaroquaux** [comment]
Very interesting!! You know these questions better than us, now, so your input is super useful. Tell us when you want a second review.

**Comment by GaelVaroquaux** [comment]
Darn, the tests errors are showing us that the strategy here is probably quite sensitive to the version of pandas.

File: dirty_cat/test/test_super_vectorizer.py
Comment (GaelVaroquaux) [inline dirty_cat/test/test_super_vectorizer.py:26]: I don't need to protect a line return if it is in the middle of enclosing parenthesis.

**Review by LilianBoulard — COMMENTED** [review]
Looks good !

File: CHANGES.rst
Comment (GaelVaroquaux) [inline CHANGES.rst:10]:
```suggestion
* Improvements to the :class:`SuperVectorizer`
```

File: dirty_cat/super_vectorizer.py
Comment (GaelVaroquaux) [inline dirty_cat/super_vectorizer.py:203]: I'm surprised that this works if X is a numpy array. I actually think that it doesn't work.

It seems that we are indeed testing properly the SuperVectorizer with numpy arrays. I'm a bit confused :)

**Review by GaelVaroquaux — COMMENTED** [review]
A few thoughts (I'm too tired to finish tonight).

**Review by GaelVaroquaux — APPROVED** [review]
LGTM. Merging. Thanks a lot!!

**Pull request merged**
